### PR TITLE
use cluster role aggregation for admin, edit, and view

### PR DESCRIPTION
**Pull request by deads2k**
Update to use the aggregated admin, edit, and view from upstream.

@openshift/sig-security 
@sub-mod fyi

/assign simo5
/assign enj

### CODE REVIEW

**Comment by deads2k** [comment]
Oh that's cool.  Because integration tests use restricted clients and those no longer have access without the aggregation controller, we don't have project level access without running controllers.

**Comment by simo5** [comment]
Am I reading it right that we now stop checking the resulting roles, but only check the additional permission we add to kubes roles ?
I think we should keep checkign the aggregated role as resulting out of the aggregation, and not just the additions we make.

**Comment by enj** [comment]
> Oh that's cool. Because integration tests use restricted clients and those no longer have access without the aggregation controller, we don't have project level access without running controllers.

I suppose you could special case the aggregation controller in the integration test master?  Seems slightly less ugly than always running controllers.

> Am I reading it right that we now stop checking the resulting roles, but only check the additional permission we add to kubes roles ?
> I think we should keep checkign the aggregated role as resulting out of the aggregation, and not just the additions we make.

The upstream coverage unit tests handle this partially.  @deads2k I assume we will need an integration test to assert the final output of all cluster roles after aggregation?

**Comment by deads2k** [comment]
> The upstream coverage unit tests handle this partially. @deads2k I assume we will need an integration test to assert the final output of all cluster roles after aggregation?

We could.  We could also choose to structure our roles to be additive.  View gets aggregated into edit and admin.  I think it probably makes more sense to do that.  In this case, I know what gets added, so rules from both admin roles up and downstream are added for coverage checks. The current unit test is still safe.



> I suppose you could special case the aggregation controller in the integration test master? Seems slightly less ugly than always running controllers.

I could.  There's a config option for it.  I don't feel very strongly.  I'll see what's easiest.

**Comment by enj** [comment]
> We could. We could also choose to structure our roles to be additive. View gets aggregated into edit and admin. I think it probably makes more sense to do that. In this case, I know what gets added, so rules from both admin roles up and downstream are added for coverage checks. The current unit test is still safe.

This handles the coverage case.  I would still like an integration test that fails when we rebase on kube and get the new additions to admin, edit and view.  From what I can tell, after this PR merges, we will no longer have a bootstrap file that contains the final, fully aggregated value for those cluster roles.

**Comment by deads2k** [comment]
> we will no longer have a bootstrap file that contains the final, fully aggregated value for those cluster roles.

You'll get the individuals, right?

File: pkg/cmd/server/bootstrappolicy/policy.go [hunk 1]
```diff
@@ -266,35 +266,12 @@ func GetOpenshiftBootstrapClusterRoles() []rbac.ClusterRole {
...
 		},
 		{
-			ObjectMeta: metav1.ObjectMeta{
-				Name: AdminRoleName,
-				Annotations: map[string]string{
-					oapi.OpenShiftDescription: "A user that has edit rights within the project and can change the project's membership.",
-				},
-			},
+			// a role for a namespace level admin.  It is `edit` plus the power to grant permissions to other users.
+			ObjectMeta: metav1.ObjectMeta{Name: "system:openshift:aggregate-to-admin", Labels: map[string]string{"rbac.authorization.k8s.io/aggregate-to-admin": "true"}},
```
Comment (soltysh) [inline pkg/cmd/server/bootstrappolicy/policy.go]: Finally, I've been waiting for this. Especially when I was trying to figure out the rights changes during 1.8 rebase.

**Comment by enj** [comment]
> You'll get the individuals, right?

Right but its the final union I care about.  @simo5 WDYT?

**Comment by deads2k** [comment]
> You'll get the individuals, right?
> Right but its the final union I care about. @simo5 WDYT?

It will be much easier to manage by splitting it into upstream and downstream.  We keep a lid on what's going in upstream.  We always separate the bootstrap file update into something rather small in the rebase, that's the place to be inspecting these.  Duplicating a role to present a whitelist is kind of silly.

**Comment by simo5** [comment]
@deads2k I do not think that double checking the full role is silly.
During rebses people are busy fixing a ton of stuff and things may slip. If you have a test that is explicitly failing you will pay attention to it.

**Comment by liggitt** [comment]
I think the diff in the upstream `system:aggregate-to-*` role is sufficient... that still prompts us to review the changes

**Comment by deads2k** [comment]
> During rebses people are busy fixing a ton of stuff and things may slip. If you have a test that is explicitly failing you will pay attention to it.

It wouldn't have any more attention paid than the diff to the role.  Less probably, since it would look like a simple fix up and most people won't even know what the resources being added do.  The diff on the list of all roles is a thing we always review and is always reviewed by someone who has a shot at knowing what they all are.

**Comment by enj** [comment]
I will write the integration test for my own sake afterwards.

Otherwise I think you will have to update the TSB bootstrap to use aggregation since they directly update admin/edit/view IIRC.

**Comment by enj** [comment]
> Otherwise I think you will have to update the TSB bootstrap to use aggregation since they directly update admin/edit/view IIRC.

@deads2k you broke SC https://bugzilla.redhat.com/show_bug.cgi?id=1535639 - `pkg/oc/bootstrap/docker/openshift/rbac.go`

**Comment by deads2k** [comment]
> Otherwise I think you will have to update the TSB bootstrap to use aggregation since they directly update admin/edit/view IIRC.
> @deads2k you broke SC https://bugzilla.redhat.com/show_bug.cgi?id=1535639 - pkg/oc/bootstrap/docker/openshift/rbac.go

Ouch.  Someone was trying to manipulate role subject to reconciliation?

**Pull request merged**